keps/sig-storage/20190204-scale-testing.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Volume Scale and Performance Testing Plan
+authors:
+  - "@msau42"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-scalability
+reviewers:
+  - "@pohly"
+  - "@saad-ali"
+  - "@wojtek-t"
+approvers:
+  - "@saad-ali"
+  - "@wojtek-t"
+editor: TBD
+creation-date: 2019-02-04
+last-updated: 2019-02-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Volume Scale and Performance Testing Plan
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+   * [Goals](#goals)
+   * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+   * [WIP Test Environment](#wip-test-environment)
+   * [Test Cases](#test-cases)
+      * [Pod Startup](#pod-startup)
+      * [Pod Teardown](#pod-teardown)
+      * [WIP: PV Binding Tests](#wip-pv-binding-tests)
+      * [WIP: PV Provisioning Tests](#wip-pv-provisioning-tests)
+      * [WIP: PV Deletion Tests](#wip-pv-deletion-tests)
+* [Graduation Criteria](#graduation-criteria)
+   * [Phase 1](#phase-1)
+   * [Phase 2](#phase-2)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+This KEP outlines a plan for testing scalability and performance of K8s storage components.
+
+## Motivation
+
+Adding storage scale and performance tests will help:
+* Understand the current scale limits of the Kuberentes storage system.
+* Set expectations (SLOs) for consumers of the Volume API.
+* Determine bottlenecks and influence which need addressing.
+
+### Goals
+
+* Measure the overhead of K8s components owned by sig-storage:
+  * K8s volume controllers
+  * Kubelet volume manager
+  * CSI sidecars
+* Stress various dimensions of volume operations to determine:
+  * Max volumes per pod
+  * Max volumes per node
+  * Max volumes per cluster
+* Test with the following volume types:
+  * EmptyDir
+  * Secret
+  * Configmap
+  * Downward API
+  * CSI mock driver
+  * TODO: Hostpath?
+  * TODO: Local?
+* Provide tests that vendors can easily run against their volume drivers.
+
+### Non-Goals
+
+* Test and measure storage provider’s drivers.
+
+## Proposal
+
+### WIP Test Environment
+
+The tests should be developed and run in sig-scalability’s test framework and infrastructure.
+
+TODO details and links
+
+### Test Portability
+
+In order to make it easier for vendors to run these tests against their volume
+drivers, the tests for persistent volumes should only use the default
+StorageClass configured for the cluster.
+
+Therefore these tests have the following prerequisites:
+
+* A Kubernetes cluster is brought up in an appropriate environment that can
+  support the storage provider.
+* The storage backend and driver is already installed and configured.
+* A default StorageClass for that driver been installed in the cluster.
+
+### Test Cases
+
+#### Pod Startup
+
+These tests should measure how long it takes to start up a pod with volumes, assuming that the volume has already been provisioned.
+
+For each volume type:
+
+* Create many pods with 1 volume each.
```

Comment:
Is that expected to start one pod after the other, waiting for it to run before creating the next, or is that a "ask for n pods at once, wait for all of them to run" test?

How many is "many"? In general, a lot of this testing is likely to hit various scalability issues. Do we want to time-box each configuration to a certain configurable duration? That's easier when starting pods one-by-one because then we can basically stop at any time. When creating n pods, n might be so large that we exceed the time limit. Then what? Looking at currently running tests at the end of the time limit is racy. Consider it a test failure?

Reply:
I believe the current scale tests create the pods in parallel.

How many needs to be figured out, as we currently don't know what our scale limits are.  I imagine that the number of pods is configurable, and we run the test with various numbers, collect the results, and see when the numbers start to degrade significantly. 

So we should have a timebox for the tests, but we'll need to do some experimentation to figure out what the limit is to fit within it. I will add a section that describes this